### PR TITLE
VAULT 15204/ldap backend config + static role resource

### PR DESCRIPTION
**Pull request by raymonstah**
Adds LDAP secret resources
* LDAP Secret Backend
* LDAP Static Role 

Also adds a LDAP container to build pipeline for acceptance tests.

More LDAP resources & docs to come in upcoming PRs.

### CODE REVIEW

File: vault/resource_ldap_secret_backend.go
Comment (raymonstah) [inline vault/resource_ldap_secret_backend.go:19]: fields pulled from https://github.com/hashicorp/vault-plugin-secrets-openldap/blob/0d09d4c5613321f936a42cafcfc9d9f5559a9e8c/path_config.go#L63

Comment (austingebauer) [reply vault/resource_ldap_secret_backend.go:19]: Thanks for being so thorough here! Like I mentioned earlier, super confusing that the LDAP secrets engine returns config fields from LDAP auth.. 

I'm thinking we can start with what's documented at https://developer.hashicorp.com/vault/api-docs/secret/ldap#configuration-management? That is _probably_ the most accurate source for the config parameters used in LDAP secrets. If we missed something, we can always expand from there. Lmk your thoughts!

File: docker-compose.yaml
Comment (austingebauer) [inline docker-compose.yaml:49]: 👍

File: vault/resource_ldap_secret_backend.go
Comment (austingebauer) [inline vault/resource_ldap_secret_backend.go:19]: I think this is missing https://developer.hashicorp.com/vault/api-docs/secret/ldap#schema?

**Review by austingebauer — APPROVED** [review]
This generally LGTM once the config fields are settled. I'll defer to TF experts on finer details 🙂

File: vault/resource_ldap_secret_backend_static_role_test.go
Comment (robmonte) [inline vault/resource_ldap_secret_backend_static_role_test.go:46]: Nit: A couple of values I've seen across the tests have some consts defined you can use. `FieldBackend`, `FieldRole`, `FieldUsername` are the ones I noticed.

File: vault/resource_ldap_secret_backend.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend.go:528]: Similar comment as above. We can make a list of API fields and loop over them as done [here](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_kubernetes_secret_backend_role.go#L200-L218)

File: vault/resource_ldap_secret_backend.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend.go:570]: We can probably consolidate the create and update methods into a single method, and use the `d.IsNewResource()` util to determine whether we are in a create or update call. An example of how the consolidation can be done is [provided here](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_kubernetes_secret_backend.go#L79-L89)

File: vault/resource_ldap_secret_backend_static_role.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend_static_role.go:26]: Suggest using constants for field names

**Review by vinay-gopalan — COMMENTED** [review]
Looking great so far! Thanks for working on this 🙏🏼 Had a couple comments/suggestions :)

File: vault/resource_ldap_secret_backend.go
Comment (raymonstah) [reply vault/resource_ldap_secret_backend.go:19]: heh, i just added it a few minutes ago.

**Comment by raymonstah** [comment]
Moving back to draft until I apply the changes suggested by @vinay-gopalan!

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend.go:314]: nit: I was thinking we could simplify all of the below `d.Set` calls to something like what we do in [resource_gcp_auth_backend](https://github.com/hashicorp/terraform-provider-vault/blob/09dc5399b590ef1e58cc00f0ca8548bbf1fcb0b5/vault/resource_gcp_auth_backend.go#L253-L265). But then I found the [resource_gcp_secret_backend sets the state](https://github.com/hashicorp/terraform-provider-vault/blob/09dc5399b590ef1e58cc00f0ca8548bbf1fcb0b5/vault/resource_gcp_secret_backend.go#L162-L166) from the [Sys.ListMounts()](https://github.com/hashicorp/terraform-provider-vault/blob/09dc5399b590ef1e58cc00f0ca8548bbf1fcb0b5/vault/resource_gcp_secret_backend.go#L147) response. If that is better maybe we could do so here? I don't know which is preferred.

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend.go:207]: I believe the recommendation is to use `d.Get` for booleans because we can't really trust that the exists response is accurate.

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend.go:446]: Same comment as above to prefer `d.Get` for booleans

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [reply vault/resource_ldap_secret_backend.go:314]: oh, woops didn't see Vinay already made this recommendation! :)

File: vault/resource_ldap_secret_backend.go
Comment (raymonstah) [reply vault/resource_ldap_secret_backend.go]: can you elaborate on this? It looks like `d.Get` just defers to `d.GetOk`. If the key is missing, then we don't set the `data` map.

Comment (fairclothjm) [reply vault/resource_ldap_secret_backend.go]: For most boolean fields, using the Get() method is preferred over GetOk(). The GetOk() evaluates the boolean and returns its value, as opposed to returning whether the boolean was set or not. This does not work in the scenario that the boolean’s value is false, as d.GetOk evaluates the boolean value and returns an ok value of false , thinking that the boolean was not set. But in fact, it was set, just with a value of false.

Comment (raymonstah) [reply vault/resource_ldap_secret_backend.go]: Oh strange behavior.. I updated the test to account for this. Thanks! I wouldn't have caught that if you didn't point it out

File: vault/resource_ldap_secret_backend_static_role_test.go
Comment (Zlaticanin) [inline vault/resource_ldap_secret_backend_static_role_test.go:32]: small nit: maybe we can set these using variables? Vinay made this suggestion for my PR as well, I think this is what we do in most of the tests.

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [reply vault/resource_ldap_secret_backend.go:19]: I believe we can remove all of the following:

```
FieldCaseSensitiveNames
FieldMaxTTL
FieldTLSMaxVersion
FieldTLSMinVersion
FieldTTL
```

And this one is labeled deprecated in the docs so maybe we should keep it?

```
FieldLength
```

File: vault/resource_ldap_secret_backend.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend.go:209]: I think we will want starttls here as well

File: vault/resource_ldap_secret_backend_static_role.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend_static_role.go:74]: nit: we could consider making this variable package-level to be reused for the create and read

File: vault/resource_ldap_secret_backend_static_role_test.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend_static_role_test.go:30]: Is this Println left over from debugging?

Comment (raymonstah) [reply vault/resource_ldap_secret_backend_static_role_test.go:30]: oops, removed 👍🏼

**Review by fairclothjm — APPROVED** [review]
LGTM! Thanks!

File: vault/resource_ldap_secret_backend_static_role_test.go
Comment (fairclothjm) [inline vault/resource_ldap_secret_backend_static_role_test.go:17]: Maybe we could add a comment on how to get the ACC tests running locally?

File: vault/resource_ldap_secret_backend.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend.go:128]: This might have to be updated if you decide to go with `path` :)

File: vault/resource_ldap_secret_backend.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend.go:133]: Related to comment above; we can avoid excluding `path` here if we decide to go this route over `mount`

**Review by vinay-gopalan — COMMENTED** [review]
Looking great! Had a comment around removing either one of the duplicate fields `mount` or `path`. Happy to answer any questions around the comments if needed. Thanks again for working on this! 🙏🏼

**Comment by vinay-gopalan** [comment]
@raymonstah could we also include docs for these resources in this PR?

File: website/docs/d/ldap_secret_backend.html.md
Comment (vinay-gopalan) [inline website/docs/d/ldap_secret_backend.html.md:1]: These will live under `website/r` since these are resources 😄

File: website/docs/d/ldap_secret_backend.html.md
Comment (vinay-gopalan) [inline website/docs/d/ldap_secret_backend.html.md:6]:
```suggestion
  Creates a LDAP secret backend for Vault.
```

File: website/docs/d/ldap_secret_backend_static_role.html.md
Comment (vinay-gopalan) [inline website/docs/d/ldap_secret_backend_static_role.html.md:6]:
```suggestion
  Creates a static role for the LDAP secret backend for Vault.
```

**Review by fairclothjm — APPROVED** [review]

File: vault/resource_ldap_secret_backend_test.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend_test.go:76]: nit:
```suggestion
				consts.FieldBindPass, consts.FieldSchema, consts.FieldDescription, consts.FieldDisableRemount),
```

File: vault/resource_ldap_secret_backend_static_role.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend_static_role.go:58]: This can be removed :)
The nuanced cases where we need to handle updates for mount are only handled by resources that are responsible for mounting secret/auth engines. Since the static role resource does not have that responsibility, this isn't needed here
 
```suggestion
```

File: vault/resource_ldap_secret_backend_static_role.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend_static_role.go:50]: Similar to comment below, this is only needed for resources that handle mounting/unmounting of secret/auth engines. Can be removed from the static role resource

```suggestion
```

File: vault/resource_ldap_secret_backend_static_role_test.go
Comment (vinay-gopalan) [inline vault/resource_ldap_secret_backend_static_role_test.go:54]: Since we removed `MustAddMountMigrationSchema` from the role resource, the `disable_remount` field is no longer part of our schema, so it can be removed from the ignore list
```suggestion
			testutil.GetImportTestStep(resourceName, false, nil, consts.FieldPath, consts.FieldRole),
```

**Review by vinay-gopalan — APPROVED** [review]
Looks great! Saw a couple things that can be removed from the static roles resource code, but should be good to merge otherwise! Thanks once again for working on this 🙏🏼

**Pull request merged**
